include/Geography.hpp

```diff
@@ -0,0 +1,199 @@
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef INCLUDE_GEOGRAPHY_H_
+#define INCLUDE_GEOGRAPHY_H_
+#include <vector>
+#include "GeographyPoint.hpp"
+
+
+namespace voltdb {
+
+class ByteBuffer;
+
+struct Geography {
+    struct Ring {
+        Ring() {}
+
+        void clear() {
+            m_points.clear();
+        }
+
+        Ring &addPoint(const GeographyPoint &aPoint) {
+            m_points.push_back(aPoint);
+            return *this;
+        }
+
+        int numPoints() const {
+            return m_points.size();
+        }
+
+        const GeographyPoint &getPoint(int idx) const {
+            return m_points[idx];
+        }
+
+        GeographyPoint &getPoint(int idx) {
+            return m_points[idx];
+        }
+
+        /**
+         * Sometimes rings need to be reversed.  We do this by reversing
+         * all but the first and last. The first and last should be equal.
+         */
+        void reverse();
+
+        bool operator==(const Ring &aOther) const {
+            return approximatelyEqual(aOther, 0.0);
+        }
+
+        bool operator!=(const Ring &aOther) const {
+            return !operator==(aOther);
+        }
+
+        bool approximatelyEqual(const Ring &rhs, double epsilon) const;
+
+        std::string toString() const;
+                       
+        void serializeTo(ByteBuffer &buffer, bool reverseit) const;
+    private:
+        typedef std::vector<GeographyPoint> PointVector;
+        typedef PointVector::iterator       PointIterator;
+        typedef PointVector::const_iterator PointConstIterator;
+        PointVector                         m_points;
+    };
+
+    Geography() {}
+
+    /**
+     * Add an already existing ring.  This is used in
+     * operator<< to add rings to Geographies.  A much
+     * more efficient location is to sue addEmptyRing()
+     * (c.f. below).
+     */
+    Geography &addRing(const Ring &aRing) {
+        m_rings.push_back(aRing);
+        return *this;
```

Comment:
Well, this is here in case one has a ring to add.  I don't use it in the tests, as there
is a more efficient way to add rings.